### ISSUE TITLE
Utc nanoseconds do not always roundtrip

### ISSUE THREAD

**Issue by crepererum** [chronotope/chrono#1289]
# Reproducer

## Code
```rust
#[test]
fn nano_roundrip() {
    let nanos = i64::MIN + 2;
    let dt = Utc.timestamp_nanos(nanos);
    let nanos2 = dt.timestamp_nanos();
    assert_eq!(nanos, nanos2);
}
```

## Expected Result
Test passes.

## Actual Result
Test panics when calling `dt.timestamp_nanos()` (so we don't even get to the actual comparison:

`value can not be represented in a timestamp with nanosecond precision.`

# Technical Background
`dt.timestamp()` is `-9223372037`, however `nanos / 1_000_000_000` is `-9223372036`, hence I conclude that this IF branch was taken:

https://github.com/chronotope/chrono/blob/b64cedc6a8758d15265702895d917045ce1eebf8/src/offset/mod.rs#L427-L430

Hence this `checked_mul` overflows:

https://github.com/chronotope/chrono/blob/b64cedc6a8758d15265702895d917045ce1eebf8/src/naive/datetime/mod.rs#L482

~My guess is that the conversion routine (first code block) isn't entirely accurate.~ The conversion is alright, I rather think when converting back we have to be more careful.

**Comment by pitdicker** [comment]
@crepererum Excellent issue report.

As far as I can tell in this code the first `checked_mul` pushes the value beyond `i64::MIN`, and the following addition should bring it back in range. Which of course doesn't work...
```rust
    pub fn timestamp_nanos(&self) -> i64 {
        self.timestamp()
            .checked_mul(1_000_000_000)
            .and_then(|ns| ns.checked_add(i64::from(self.timestamp_subsec_nanos())))
            .expect("value can not be represented in a timestamp with nanosecond precision.")
    }
```

For negative timestamps we should instead add `1` to the timestamp, do the multiply and addition, and subtract `1_000_000_000`.

@crepererum Interested in making a PR?

**Comment by crepererum** [comment]
Fixed by #1294 and released in [0.4.31](https://github.com/chronotope/chrono/releases/tag/v0.4.31).
